ReText/singleapplication.py

```diff
@@ -0,0 +1,167 @@
+			return
+
+		self._localSockets[localSocket] += localSocket.readAll()
+		data = self._localSockets[localSocket]
+		if len(data) > 4:
+			# First 4bytes is an native Integer.
+			dataSize = struct.unpack("@I", data[:4])[0]
+			receivedDataSize = len(data) - 4
+			if receivedDataSize < dataSize:
+				return
+
+			self.receivedMessage.emit(bytes(data[4:]))
+
+			# Remove the command socket
+			del self._localSockets[localSocket]
+			localSocket.deleteLater()
+
+	def _onServerNewConnection(self):
+		while self._server.hasPendingConnections():
+			localSocket = self._server.nextPendingConnection()
+			self._localSockets[localSocket] = b""
+			localSocket.readyRead.connect(functools.partial(
+				self._onLocalSocketReadyRead, localSocket))
+
+	def start(self):
+		# Ensure we run only one application
+		isAnotherRunning = False
+		if not self._sharedMemory.isAttached():
+			# If share memory is attachable, that means there have another
+			# running application
+			self._systemSemaphore.acquire()
+			try:
+				isAnotherRunning = self._sharedMemory.attach()
+				if isAnotherRunning:
+					self._sharedMemory.detach()
+			finally:
+				self._systemSemaphore.release()
+
+		if not isAnotherRunning:
+			# If we already attached by previous attach test, we will create
+			# the share memory.
+			self._systemSemaphore.acquire()
+			try:
+				isAnotherRunning = not self._sharedMemory.create(1, QSharedMemory.ReadWrite)
+			finally:
+				self._systemSemaphore.release()
+
+			# If there something happened during create procedure lead us can't
+			# create one, we should detach that share memory.
+			if isAnotherRunning:
+				self._systemSemaphore.acquire()
+				try:
+					if self._sharedMemory.isAttached():
+						self._sharedMemory.detach()
+				finally:
+					self._systemSemaphore.release();
+
+		# Fine, now we could take different action on different situation.
+		if not isAnotherRunning:
+			self._mode = self.Server
+			self._server = QLocalServer(self)
+			self._server.newConnection.connect(self._onServerNewConnection)
+
+			if not self._server.listen(self._name):
+				# Failed to listen, is there have another application crashed
+				# without normally shutdown it's server?
+				#
+				# We try to remove the old dancing server and restart a new
+				# server.
+				self._server.removeServer(self._name)
+				if not self._server.listen(self._name):
+					raise RuntimeError("Local server failed to listen on '%s'" % self._name)
+		else:
+			# Detach immediately if create failed.
+			#
+			# WARNING: On windows os, seems if we don't detach the share memory
+			# after failed to create, another ReText can't create the share
+			# memory after original instance exitted while there still have
+			# ReText as client.
+
+			self._mode = self.Client
+			self._client = QLocalSocket(self)
+			self._client.connectToServer(self._name)
+
+	def sendMessage(self, message):
+		# Only accept bytes message
+		assert(type(message) == bytes)
```

Comment:
It's really needed.
I will change to `isinstance()` and raise TypeError if it failed.